context.go

```diff
@@ -0,0 +1,43 @@
+package stats
+
+import "context"
+
+func ContextWithTags(ctx context.Context, tags ...Tag) context.Context {
+	if x := getTagSlice(ctx); x != nil {
+		*x = append(*x, tags...)
+		return ctx
```

Comment:
I think we should pass the tags by value and make a new slice and a new context every time here, otherwise the memory that is implicitly shared will be a fertile ground for bugs that are hard to track down.

For example:
```go
return context.WithValue(ctx, contextKeyReqTags, append(ContextTags(ctx), tags...))
```

Reply:
I'm on board with creating a new context each time, but the pointer to slice is fundamental to this feature.  Otherwise no child contexts can update the reference and the root httpstats Handler can't get the tags from the child handlers.

Reply:
I think once we add some sync access to this key, it should be fine.  We're not exporting anything besides the functions so it will "just work" to the consumers.  If people start changing things in `context.go`, that's when we'll start to have issues.  Otherwise, I think we'll be alright.

Maybe we make a copy of the original slice just to be extra safe that nothing outside of this has a reference to the original slice address.  Callers might have access to the slice contents, but they won't be able to access the slice itself.

Reply:
OK, I think this model has different semantics than the way context functions named `With*` usually work.

How about having two functions then?

- `WithContextTags` is the immutable form, always returns a new context
- `AddContextTags` is the mutable form, returns nothing, and always modifies the context passed as argument (panic or do nothing if the context has no tags?)

Splitting the two use cases makes each API harder to use wrong.

Reply:
Updated the API to match the above ^